File: backend-dotnet/src/Walos.Application/Security/PasswordPolicy.cs
using Walos.Domain.Exceptions;
using System.Text.RegularExpressions;

namespace Walos.Application.Security;

public static class PasswordPolicy
{
    public const int MinimumLength = 8;
    private const string InvalidAsciiPattern = @"[^\x21-\x7E]";
    private const string UppercasePattern = @"[A-Z]";
    private const string LowercasePattern = @"[a-z]";
    private const string DigitPattern = @"[0-9]";
    private const string SpecialPattern = @"[^A-Za-z0-9]";

    public static void Validate(string? password)
    {
        if (string.IsNullOrWhiteSpace(password) || password.Length < MinimumLength)
            throw new ValidationException("La contraseña debe tener al menos 8 caracteres");

        if (Regex.IsMatch(password, InvalidAsciiPattern, RegexOptions.CultureInvariant)
            || !Regex.IsMatch(password, UppercasePattern, RegexOptions.CultureInvariant)
            || !Regex.IsMatch(password, LowercasePattern, RegexOptions.CultureInvariant)
            || !Regex.IsMatch(password, DigitPattern, RegexOptions.CultureInvariant)
            || !Regex.IsMatch(password, SpecialPattern, RegexOptions.CultureInvariant))
        {
            throw new ValidationException(
                "La contraseña debe usar caracteres ASCII visibles e incluir mayúscula, minúscula, número y carácter especial, sin espacios");
        }
    }
}

File: frontend/src/utils/passwordPolicy.js
export const validatePassword = (password) => {
  if (!password || password.length < 8) return 'La contraseña debe tener al menos 8 caracteres';
  if (/[^\x21-\x7E]/.test(password)) return 'Usá solo caracteres ASCII visibles, sin espacios';
  if (!/[A-Z]/.test(password) || !/[a-z]/.test(password) || !/[0-9]/.test(password) || !/[^A-Za-z0-9]/.test(password)) {
    return 'Incluí mayúscula, minúscula, número y carácter especial';
  }
  return null;
};
